Implement filesystem quota for overlay2 over ext4 if it support

coolljt0725 opened the pull request:
- fixes https://github.com/moby/moby/issues/29364

Signed-off-by: Lei Jitang <leijitang@huawei.com>

/cc @dmcgowan  @amir73il 

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
This pr adds support for --storage-opt size=xx to overlay2 over
ext4 if kernel support and mount with `-o prjquota` option.
**- How I did it**

**- How to verify it**

First we need enable the quota during mkfs with
```
mkfs.ext4 -O quota,project /dev/mapper/vg--docker-ext4
```
or you can update the old filsystem with
```
tune2fs -O quota,project /dev/mapper/vg--docker-ext4
```

Then mount the disk with `-o prjquota`
```
/dev/mapper/vg--docker-ext4 on /mnt type ext4 (rw,relatime,seclabel,prjquota,data=ordered)
```

Then start the docker and use `-g` to specify the root to `/mnt`.

Run a container and check whether the quota works.
```
$ docker run -ti --storage-opt size=10M busybox
/ # df -h
Filesystem                Size      Used Available Use% Mounted on
overlay                  10.0M     24.0K     10.0M   0% /
tmpfs                     7.8G         0      7.8G   0% /dev
tmpfs                     7.8G         0      7.8G   0% /sys/fs/cgroup
/dev/mapper/vg--docker-ext4
                         29.4G     82.4M     27.8G   0% /etc/resolv.conf
/dev/mapper/vg--docker-ext4
                         29.4G     82.4M     27.8G   0% /etc/hostname
/dev/mapper/vg--docker-ext4
                         29.4G     82.4M     27.8G   0% /etc/hosts
shm                      64.0M         0     64.0M   0% /dev/shm
tmpfs                     7.8G         0      7.8G   0% /proc/kcore
tmpfs                     7.8G         0      7.8G   0% /proc/sched_debug
/ # dd if=/dev/zero of=/home/img bs=2M count=6 && sync
dm-3: write failed, project block limit reached.
dd: writing '/home/img': Disk quota exceeded
6+0 records in
4+1 records out
10444800 bytes (10.0MB) copied, 0.022292 seconds, 446.8MB/s
/ # df -h
Filesystem                Size      Used Available Use% Mounted on
overlay                  10.0M     10.0M         0 100% /
tmpfs                     7.8G         0      7.8G   0% /dev
tmpfs                     7.8G         0      7.8G   0% /sys/fs/cgroup
/dev/mapper/vg--docker-ext4
                         29.4G     92.4M     27.8G   0% /etc/resolv.conf
/dev/mapper/vg--docker-ext4
                         29.4G     92.4M     27.8G   0% /etc/hostname
/dev/mapper/vg--docker-ext4
                         29.4G     92.4M     27.8G   0% /etc/hosts
shm                      64.0M         0     64.0M   0% /dev/shm
tmpfs                     7.8G         0      7.8G   0% /proc/kcore
tmpfs                     7.8G         0      7.8G   0% /proc/sched_debug

```
**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

AkihiroSuda commented on daemon/graphdriver/quota/projectquota.go:
```diff
@@ -38,15 +38,31 @@ struct fsxattr {
 #endif
-#ifndef XFS_PROJ_QUOTA
-#define XFS_PROJ_QUOTA	2
+#ifndef PROJ_QUOTA
```
why?

coolljt0725 replied:
because the quota is not restrict on xfs

amir73il replied:
That ifdef was added to fix build issues on some systems. Not sure you can remove or change it

AkihiroSuda commented on daemon/graphdriver/quota/projectquota.go:
```diff
@@ -114,6 +130,17 @@ func NewControl(basePath string) (*Control, error) {
+		if err != nil {
+			logrus.Debugf("Get qutoa stat failed with: %v", err)
+		}
+		return nil, fmt.Errorf("quota doesn't support on your system", err)
```
doesn't support -> isn't supported

AkihiroSuda commented on daemon/graphdriver/quota/projectquota.go:
```diff
@@ -194,6 +221,24 @@ func setProjectQuota(backingFsBlockDev string, projectID uint32, quota Quota) er
 }
 
+// getQuotaStat - get the quota stat
+// return 2 means quota is on
```
Let's avoid using magic numbers

coolljt0725 replied:
I just take this https://kernel.googlesource.com/pub/scm/utils/quota/quota-tools/+/master/quotasys.c#1123  as a reference

coolljt0725 committed:
Implement filesystem quota for overlay2 over ext4 if it support

This pr adds support for --storage-opt size=xx to overlay2 over
ext4 if kernel support and mount with `-o prjquota` option.

- how to verify it

First we need enable the quota during mkfs with
```
mkfs.ext4 -O quota,project /dev/mapper/vg--docker-ext4
```
or you can update the old filsystem with
```
tune2fs -O quota,project /dev/mapper/vg--docker-ext4
```

Then mount the disk with '-o prjquota'
```
/dev/mapper/vg--docker-ext4 on /mnt type ext4 (rw,relatime,seclabel,prjquota,data=ordered)
```

Then start the docker and use `-g` to specify the root to `/mnt`.

Run a container and check whether the quota works.
```
$ docker run -ti --storage-opt size=10M busybox
/ # df -h
Filesystem                Size      Used Available Use% Mounted on
overlay                  10.0M     24.0K     10.0M   0% /
tmpfs                     7.8G         0      7.8G   0% /dev
tmpfs                     7.8G         0      7.8G   0% /sys/fs/cgroup
/dev/mapper/vg--docker-ext4
                         29.4G     82.4M     27.8G   0% /etc/resolv.conf
/dev/mapper/vg--docker-ext4
                         29.4G     82.4M     27.8G   0% /etc/hostname
/dev/mapper/vg--docker-ext4
                         29.4G     82.4M     27.8G   0% /etc/hosts
shm                      64.0M         0     64.0M   0% /dev/shm
tmpfs                     7.8G         0      7.8G   0% /proc/kcore
tmpfs                     7.8G         0      7.8G   0% /proc/sched_debug
/ # dd if=/dev/zero of=/home/img bs=2M count=6 && sync
dm-3: write failed, project block limit reached.
dd: writing '/home/img': Disk quota exceeded
6+0 records in
4+1 records out
10444800 bytes (10.0MB) copied, 0.022292 seconds, 446.8MB/s
/ # df -h
Filesystem                Size      Used Available Use% Mounted on
overlay                  10.0M     10.0M         0 100% /
tmpfs                     7.8G         0      7.8G   0% /dev
tmpfs                     7.8G         0      7.8G   0% /sys/fs/cgroup
/dev/mapper/vg--docker-ext4
                         29.4G     92.4M     27.8G   0% /etc/resolv.conf
/dev/mapper/vg--docker-ext4
                         29.4G     92.4M     27.8G   0% /etc/hostname
/dev/mapper/vg--docker-ext4
                         29.4G     92.4M     27.8G   0% /etc/hosts
shm                      64.0M         0     64.0M   0% /dev/shm
tmpfs                     7.8G         0      7.8G   0% /proc/kcore
tmpfs                     7.8G         0      7.8G   0% /proc/sched_debug
/ #

```

Signed-off-by: Lei Jitang <leijitang@huawei.com>

amir73il commented on daemon/graphdriver/quota/projectquota.go:
```diff
@@ -194,6 +221,24 @@ func setProjectQuota(backingFsBlockDev string, projectID uint32, quota Quota) er
 
+// getQuotaStat - get the quota stat
+// return 2 means quota is on
+func getQuotaStat(backingFsBlockDev string) (int, error) {
```
@coolljt0725 getQuotaStat() is a nice addition, but it has nothing to do with adding ext4 project quota support, does it? The code for xfs project quota could just as well not be really setting project quota limits.
So if you wish to add this check, please do that in a separate commit.

This check may be important for both ext4/xfs project quota, however! it is not enough to verify that ext4 project quota can be applied to overlayfs upper layer.

There is a bug in overlayfs since kernel v4.8 that will cause overlayfs to ignore ext4 project quota limits.
The bug is fixed in kernel v4.12 and has been applied to stable kernel 4.9.y (see https://marc.info/?l=linux-unionfs&m=150095448912473&w=2), but ithe bug may very well be on deployed systems (e.g. Ubuntu 16.10).

As far as I know, the only way to check if the bug exists is to set a test hard quota limit (either user/group/project) on overlayfs upper layer and try to write beyond the limit. And if you are going to implement that test, maybe there is no need to bother with the weaker getQuotaStat() test ???

kolyshkin replied:
@coolljt0725 I second @amir73il opinion that it's better to have a run-time check if project quota is working. Please let me know if you are going to work on this.

amir73il reviewed (changes requested):
I would like to see the "How I verified" section and specifically would like to see this verified in kernel v4.8 with an overlayfs bug that prevents ext4 project quota from taking effect

amir73il commented:
cc @rhvgoyal

coolljt0725 commented on daemon/graphdriver/quota/projectquota.go:
```diff
@@ -194,6 +221,24 @@ func setProjectQuota(backingFsBlockDev string, projectID uint32, quota Quota) er
 
+// getQuotaStat - get the quota stat
+// return 2 means quota is on
+func getQuotaStat(backingFsBlockDev string) (int, error) {
```
@amir73il  Add `getQuotaStat` is to check whether the system  support ext4 quota. I didn't know much about that bug you mentioned, so if that bug exist,  `getQuotaStat` still return `2`?

amir73il replied:
Yes, it will still return 2. 2 means quota is enforced,  but root user can bypass enforcement and that is exactly what happens with overlayfs over ext4 since kernel v4.8, because overlayfs writes to upper layer with credentials of the user that mounted the overlay (usually root user).

coolljt0725 commented:
I tested this on `Kernel Version: 4.11.8-300.fc26.x86_64`, the root dir of docker has mount a lvm with `prjquota` option
```
/dev/mapper/vg--docker-ext4 on /mnt/ext4 type ext4 (rw,relatime,seclabel,prjquota,data=ordered)
```

```
[lei@localhost docker]$ docker run -ti --storage-opt size=15g busybox
/ # df -h
Filesystem                Size      Used Available Use% Mounted on
overlay                  15.0G     24.0K     15.0G   0% /
tmpfs                    64.0M         0     64.0M   0% /dev
tmpfs                     1.9G         0      1.9G   0% /sys/fs/cgroup
/dev/mapper/vg--docker-ext4
                         19.6G     46.3M     18.5G   0% /etc/resolv.conf
/dev/mapper/vg--docker-ext4
                         19.6G     46.3M     18.5G   0% /etc/hostname
/dev/mapper/vg--docker-ext4
                         19.6G     46.3M     18.5G   0% /etc/hosts
shm                      64.0M         0     64.0M   0% /dev/shm
tmpfs                    64.0M         0     64.0M   0% /proc/kcore
tmpfs                    64.0M         0     64.0M   0% /proc/latency_stats
tmpfs                    64.0M         0     64.0M   0% /proc/timer_list
tmpfs                    64.0M         0     64.0M   0% /proc/sched_debug
tmpfs                     1.9G         0      1.9G   0% /sys/firmware
/ #

```

amir73il commented:
@coolljt0725 please edit PR description with "how to verify it". The thing with the overlayfs bug is that you will see quota enforced if you only check with df -h, but if you try to actually write a file larger than quota, you should be able to do that despite of the quota enforcement.
Kernel v4.11 should have the bug fixed. Kernel v4.8.x not fixed AFAIK, but I don't follow Ubuntu stable fixes. I should mention that I never experienced that bug myself, it was reported by someone else, I only pointed out that the bug exists since kernel v4.8 and fix should be backported.

coolljt0725 commented:
@amir73il  thank you for your response. I will add some description with `how to verify it`

Do you have any test suit to test the quota function when you work on https://github.com/moby/moby/pull/24771 ? If you have , we can use that suit to test the quota function.

amir73il commented:
Sorry I don't have any test suite. I tested exactly as you tested. I only metion the bug because it was brought to my attention. Try to see if you can exceed quota limit by hardlinking a large file. Btw if docker daemon drop CAP SYS RESOURCE before mouting overlay then the bug is irrelevant.

coolljt0725 commented:
I have update the implementation for ext4, use `QCMD(Q_SETQUOTA, PRJQUOTA)` instead of `QCMD(Q_XSETQLIM, PRJQUOTA)`  to set quota for ext4 because for some backported kernel, they may not support `QCMD(Q_XSETQLIM, PRJQUOTA)`. This implementation take  https://git.kernel.org/pub/scm/utils/quota/quota-tools.git/tree/setquota.c as a reference.

And also the `how to verify it` section is updated.

coolljt0725 commented:
Add a commit to add a lock to protect the access to `projectID`

coolljt0725 commented:
ping @amir73il  @dmcgowan  @rhvgoyal

amir73il reviewed (commented):
Although no code uses GetQuota() at the moment and more importantly, no code unsets quota, it would be nice to use the lock in GetQuota() as well to make future implementation more robust.

amir73il commented on daemon/daemon_unix.go:
```diff
@@ -576,6 +576,22 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
+		_, exist := hostConfig.StorageOpt["size"]
+		if exist {
+			if hostConfig.Privileged {
+				warnings = append(warnings, "filesystem quota for overlay2 over ext4 or xfs can't take affect with privileged container")
```
Actually, that is not true for xfs. xfs does not use the generic vfs quota enforcement and enforces quotas also for privileged user (and for user with CAP_SYS_RESOURCE)

amir73il commented on daemon/graphdriver/quota/projectquota.go:
```diff
@@ -194,6 +247,49 @@ func setProjectQuota(backingFsBlockDev string, projectID uint32, quota Quota) er
+}
+
+// SetProjectQuota - set the quota for project id on ext4 block device
+func (q *Ext4Quota) SetProjectQuota(backingFsBlockDev string, projectID uint32, quota Quota) error {
```
Am I wrong to assume that this is not really the "Ext4Quota" implementation, but the new generic "VfsQuota" implementation, which can also be applied to other fs that may support project quota in the future?
Am I wrong to assume that since kernel version v4.5, both xfs and ext4 supports that same unified API?
Am I wrong to assume that if you need to use  Q_SETPQUOTA, you also need to use Q_GETPQUOTA and Q_GETPQSTAT for ext4?
If I am not wrong, I suggest to determine the QuotafileOps to use by kernel version (either XFS legacy or VFS ops) and to include also GetProjectQuota() and GetProjectQuotaStat() in the QuotafileOps struct.

coolljt0725 replied:
@amir73il  Thank you for your response and sorry for the late response :)
>　Am I wrong to assume that this is not really the "Ext4Quota" implementation, but the new generic "VfsQuota" implementation, which can also be applied to other fs that may support project quota in the future?

It's true that is `VfsQuota`, I named it to `Ext4Quota` because it's only used for ext4 at the moment, I'm fine to rename it to `VfsQuota`

>Am I wrong to assume that since kernel version v4.5, both xfs and ext4 supports that same unified API?
Am I wrong to assume that if you need to use Q_SETPQUOTA, you also need to use Q_GETPQUOTA and Q_GETPQSTAT for ext4?
If I am not wrong, I suggest to determine the QuotafileOps to use by kernel version (either XFS legacy or VFS ops) and to include also GetProjectQuota() and GetProjectQuotaStat() in the QuotafileOps struct.

The main reason I  implemented like this is we have a backported kernel (3.10 with some ext4 project quota)，the API for Ext4 and XFS are not the same unified API.
 Because of the backported kernel, determining the QuotafileOpts to use by kernel version is a hard thing. 
For `GetQuotaStat`,  from what I know, the `vfsQuota` hasn't implemented it yet(https://github.com/torvalds/linux/blob/master/fs/quota/quota.c#L718), so we still need to use `Q_XGETQSTAT`.  

So how about use `Q_GETQUOTA` and `Q_SETQUOTA` in `QuotafileOps` for ext4 and use `Q_XSETQLIM` and `Q_XGETQUOTA` for xfs?

amir73il replied:
Ok, but for future code sake let's use Q_X.. for xfsQuota and Q_... For VfsQuota and try to avoid checking for ext4 explicitly when choosing VfsQuota. Thanks

coolljt0725 committed:
Add lock to project quota

Signed-off-by: Lei Jitang <leijitang@huawei.com>

coolljt0725 commented:
updated

amir73il reviewed (commented):
Code reuse between legacy and vfs implementations could be better otherwise LGTM

thaJeztah commented:
ping @dmcgowan PTAL

kolyshkin commented:
@coolljt0725 so what is the reason you prefer to use ext4 over xfs?

kolyshkin commented:
Some relevant quotes from RHEL docs

RHEL 7.1: OverlayFS is supported as a Technology Preview. There are currently two restrictions:
It is recommended to use ext4 as the lower file system; the use of xfs and gfs2 file systems is not supported.

RHEL 7.2, 7.3, 7.4: OverlayFS remains a Technology Preview in Red Hat Enterprise Linux 7.2 under most circumstances. Only XFS is currently supported for use as a lower layer file system.

coolljt0725 commented:
@kolyshkin  The ext4 is the only filesystem we support  on our production environment(ext4 is much more steady and we encountered several bugs on xfs that's why we abandoned  it ) and we have backported the ext4 project quota from the mainline.

kolyshkin commented:
I think this PR needs some work:
1. address @amir73il's review comment about having a run-time check for working project quota https://github.com/moby/moby/pull/34364/commits/cd5e1162f23a8fd951c82244fc4438fc782c41c1#r130996779
2. Perhaps GetQuota/SetQuota can be unified?

What are your thoughts @coolljt0725 ?

bklau commented:
@thaJeztah @coolljt0725  The Official Docker documentation gives the impression that this PR is already merged and  supported:
https://docs.docker.com/engine/userguide/storagedriver/selectadriver/#docker-ce
 Q:  This can't be true, is it?. Pls clarify.

thaJeztah commented:
@bklau can you elaborate why the docs suggest this PR is merged already?

bklau commented:
@thaJeztah Read this link : https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/#prerequisites
"

> The following backing filesystems are supported:
> ext4 (RHEL 7.1 only)
> xfs (RHEL 7.2 and higher), but only with d_type=true enabled. Use xfs_info to verify that the ftype option is set to 1. To format an xfs filesystem correctly, use the flag -n ftype=1
"
Q: So without this merge, the overlay2 with ext4 would just worked if I have RHEL 7.1?  Is there anything special with  RHEL 7.1 that makes this worked without this merge?. That's what raced thru my mind.

bklau commented:
@thaJeztah  Is my  comment on  the ext4 backing support in overlay2 correct?. If not, pls clarify. Thx.

thaJeztah commented:
Ping @coolljt0725 what's the status on this one? I see some comments from @kolyshkin that were not addressed yet, correct? https://github.com/moby/moby/pull/34364#issuecomment-335974297

thaJeztah commented:
Also; we'd need integration tests for this

coolljt0725 commented:
@thaJeztah  sorry, I'm almost forget this pr. let me close this at the moment and re-open it when it's ready.

katsar0v commented:
Is there any update on this?

thaJeztah commented:
@katsar0v no update, afaik, but if someone wants to work on this, and carry this change, that would be appreciated.
